### PR TITLE
Add http patch support for annotations

### LINKED ISSUE #12546: Support PATCH verb in annotations http api
### What Grafana version are you using?
v5.3.0

### What did you do?
Calling a PUT for /api/annotations/ to update an existing annotation

### What was the expected result?
The values passed in would be added to the annotation, without loss of existing data

### What happened instead?
Fields not passed in are nulled out

An example request looks like the following:
`curl -vv -X PUT -H "Content-Type: application/json"  -H "Authorization: Bearer eyJrIjoiTj...iOjF9" -d '{"text":"Annotation Description"}' http://localhost:3000/api/annotations/1`

Before the `PUT`, the annotation contains several other data (e.g., tags, time, etc.). When updating the annotation with `"text":"Annotation Description"`, fields not passed in are nulled out.

### PR DESCRIPTION
**Pull request by SamuelToh**
Fixes #12546 

I'm happy to send in another pull request for the documentation side of things once this is accepted.

### CODE REVIEW

**Comment by SamuelToh** [comment]
Any chance someone can retrigger the CI? I'm not sure what it is failing on, not much information was provided by the failure.

Running `go test ./pkg/...` seem to work fine on my machine.
```

root@samuelt-ThinkPad-T460s:~/go/src/github.com/grafana/grafana# go test ./pkg/...
ok  	github.com/grafana/grafana/pkg/api	0.159s
?   	github.com/grafana/grafana/pkg/api/avatar	[no test files]
ok  	github.com/grafana/grafana/pkg/api/dtos	0.005s
?   	github.com/grafana/grafana/pkg/api/live	[no test files]
ok  	github.com/grafana/grafana/pkg/api/pluginproxy	0.197s
ok  	github.com/grafana/grafana/pkg/api/routing	0.014s
?   	github.com/grafana/grafana/pkg/api/static	[no test files]

```

**Comment by SamuelToh** [comment]
Also just in case someone is reviewing this. The `PATCH` interface is almost the same as `UPDATE`.

You can basically `PATCH` almost any properties of an annotation, including timeEnd. Only thing you can't `patch` was going from a "isRegion" annotation to non region ones, as it involves `deleting` the 'right hand side' annotation. Not sure we should allow this or not, as `patching` generally means we only update the annotation, not involving deletion.

**Review by marefr — CHANGES_REQUESTED** [review]
This is looking pretty good :+1:. Only reviewed the code, not tested it out. 

Please fix according to comments.
Regarding documentation we would like that to be updated in this PR as well.

You also have a failing build due to go formatting issues. Please run `go fmt pkg/api/annotations*` to fix the problems.

File: pkg/api/annotations.go [hunk 1]
```diff
@@ -210,6 +210,61 @@ func UpdateAnnotation(c *m.ReqContext, cmd dtos.UpdateAnnotationsCmd) Response {
...
+func PatchAnnotation(c *m.ReqContext, cmd dtos.PatchAnnotationsCmd) Response {
+	annotationID := c.ParamsInt64(":annotationId")
+
+	repo := annotations.GetRepository()
+
+	if resp := canSave(c, repo, annotationID); resp != nil {
+		return resp
+	}
+
+	items, _ := repo.Find(&annotations.ItemQuery{AnnotationId: annotationID, OrgId: c.OrgId})
```
[the remark this answers is not in the document]
Comment (SamuelToh) [reply pkg/api/annotations.go]: Thanks mate will update according tonight and send for review again

[the remark this answers is not in the document]
Comment (SamuelToh) [reply pkg/api/annotations.go]: @marefr All good now.

- Added doc describing new endpoint.
- Addressed review comment about the len 0 check.
- Rectified `fmt`'s formatting complain.

[the remark this answers is not in the document]
Comment (marefr) [reply pkg/api/annotations.go]: Seems you still have some go formatting issues, please run `go fmt pkg/api/dtos/annotations.go` to resolve those and then push the changes. Thanks

**Review by marefr — CHANGES_REQUESTED** [review]
Starting to look pretty solid. Would like some minor changes (see comments).

File: docs/sources/http_api/annotations.md [hunk 2]
```diff
@@ -180,6 +183,28 @@ Content-Type: application/json
 }
 ```
 
+## Update Annotation
+
+`PATCH /api/annotations/:id`
+
+Updates one or more properties of an annotation that matches the specified id.
```
[the remark this answers is not in the document]
Comment (SamuelToh) [reply docs/sources/http_api/annotations.md]: Fixed. Added a paragraph to describe what can be updated and what not.

File: docs/sources/http_api/annotations.md [hunk 3]
```diff
@@ -180,6 +183,28 @@ Content-Type: application/json
...
+
+**Example Request**:
```
[the remark this answers is not in the document]
Comment (SamuelToh) [reply docs/sources/http_api/annotations.md]: fixed

File: pkg/api/annotations.go [hunk 4]
```diff
@@ -210,6 +210,65 @@ func UpdateAnnotation(c *m.ReqContext, cmd dtos.UpdateAnnotationsCmd) Response {
...
+	repo := annotations.GetRepository()
+
+	if resp := canSave(c, repo, annotationID); resp != nil {
+		return resp
+	}
+
+	items, err := repo.Find(&annotations.ItemQuery{AnnotationId: annotationID, OrgId: c.OrgId})
+
+	if err != nil || len(items) == 0 {
+		return Error(500, "Could not find annotation to update", err)
```
[the remark this answers is not in the document]
Comment (SamuelToh) [reply pkg/api/annotations.go]: fixed.

File: pkg/api/annotations.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SamuelToh) [reply pkg/api/annotations.go]: @marefr  Addressed all comments. I think i'm all good now.

**Review by marefr — APPROVED** [review]
Now I have done some manual testing as well and everything works as expected.

Awesome! Great work!

**Comment by marefr** [comment]
Thank you for contributing to Grafana!

**Pull request merged**